usine.py

```diff
@@ -1,23 +1,349 @@
+
+        return {'username': user, 'hostname': host, 'port': port}
+
+    def execute(self, cmd, **kwargs):
+        channel = self._transport.open_session()
+        prefix = self.prefix
+        if self.cd:
+            cmd = f'cd {self.cd}; {cmd}'
+        if self.env:
+            env = ' '.join(f'{k}={v}' for k, v in self.env.items())
+            prefix = f'{prefix} {env}'
+        cmd = self.format(prefix + " sh -c '" + cmd + "'")
+        print(gray(cmd))
+        channel.exec_command(cmd)
+        stdout = channel.makefile('r', -1)
+        stderr = channel.makefile_stderr('r', -1)
+        buf = b''
+        while True:
+            data = stdout.read(1)
+            if not data:
+                if buf:
+                    sys.stdout.write(buf.decode())
+                break
+            buf += data
+            if buf.endswith(b'\n'):
+                sys.stdout.write(buf.decode())
+                buf = b''
+        ret = Status(stdout, stderr, channel.recv_exit_status())
+        channel.close()
+        if ret.code:
+            red(ret.stderr)
+            sys.exit(ret.code)
+        return ret
+
+    def format(self, tpl):
+        return self.formatter.vformat(tpl, None, self.context)
+
+    @property
+    def sftp(self):
+        if not self._sftp:
+            self._sftp = self._client.open_sftp()
+        return self._sftp
+
+
+@contextmanager
+def connect(*args, **kwargs):
+    global client
+    client = Client(*args, **kwargs)
+    yield client
+    client.close()
 
 
 def run(cmd):
-    if connection.context['sudo']:
-        cmd = f'sudo {cmd}'
-    connection.execute(cmd)
+    return client.execute(cmd)
+
+
+def exists(path):
+    return bool(run(f'if [ -e "{path}" ]; then echo 1; fi').stdout)
+
+
+@formattable
+def mkdir(path, parents=True, mode=None):
+    return run('mkdir {parents:bool} {mode:equal} {path}')
+
+
+@formattable
+def chown(mode, path, recursive=True, preserve_root=True):
+    return run('chown {recursive:bool} {mode} {path}')
+
+
+@formattable
+def ls(path, all=True, human_readable=True, size=True, list=True):
+    return run('ls {all:bool} {human_readable:bool} {size:bool} {list:initial}'
+               ' {path}')
+
+
+def mv(src, dest):
+    return run(f'mv {src} {dest}')
+
+
+def put(local, remote, owner=None):
+    bar = ProgressBar(prefix=f'{local} => {remote}')
+    tmp = str(Path('/tmp') / md5(remote.encode()).hexdigest())
+    func = client.sftp.putfo if hasattr(local, 'read') else client.sftp.put
+    func(local, tmp, lambda done, total: bar.update(done=done, total=total))
+    mv(tmp, remote)
+    if owner:
+        chown(owner, remote)
+
+
+def get(remote, local):
+    bar = ProgressBar(prefix=f'{remote} => {local}')
+    func = client.sftp.getfo if hasattr(local, 'write') else client.sftp.get
+    func(remote, local, lambda done, total: bar.update(done=done, total=total))
+
+
+@contextmanager
+@formattable
+def sudo(set_home=True, preserve_env=True, user=None, login=None):
```

Comment:
Could we also use it as a decorator?
```
@cli
@sudo
def my_cmd():
   ...
```

Reply:
To be consistent we would have to allow that for `cd` and `env` too?